### DISCUSSION TITLE
Retrieve Evaluation Group table

### DISCUSSION THREAD

**Discussion by JohannesScheyerle** [MPh-py/MPh#184] [General]
Hey,

my general request is, how do i directly address simulated data from an mph file in python. 

My case is that I set up and simulate everything in Comsol in box. In the end, I have an evaluation group including a general transformation over three surface averages. I can then plot in box in Comsol the results from this evaluation group in a table graph. 

But I would love to get this data that fuels the plot in Comsol (Table from Evaluation group) out in python. My workaround at this time is to export the evaluation group table as .txt and import this in python. 

But is there a way via mph to directly get the data from an evaluation group out of an .mph file into python? This would be very convenient for me. If yes, how?

Best,
Johannes

**Comment by john-hen** [comment]
> But is there a way via mph to directly get the data from an evaluation group out of an .mph file into python? This would be very convenient for me.

Depends what you mean by "directly". There's no [API] method for accessing the data in a Comsol table. The API methods are the "convenience" functions that MPh provides. So if convenience is what you're looking for, then no.

It's certainly possible though, given that you can [access the full Comsol API]. I wouldn't call the Comsol API "convenient", which is why we have our own API. It's just easier to use. But your answer is to be found "somewhere" in the Comsol [Programming Reference Manual]. (Which is "only" 1182 pages long.) Once you know how to do it in Java, doing the same thing in Python is straightforward. See also the pinned discussion.

There is an API method for evaluating Comsol expressions: [`Model.evaluate()`]. For me, personally, that was all I ever needed to get the data from Comsol into Python. Though you may have to define the occasional "coupling operator" just so you can evaluate an expression that returns the data you need. That is, whatever the evaluation groups do in your particular use case.

[API]: https://mph.readthedocs.io/en/1.2/api.html
[Programming Reference Manual]: https://doc.comsol.com/6.2/doc/com.comsol.help.comsol/COMSOL_ProgrammingReferenceManual.pdf
[access the full Comsol API]: https://mph.readthedocs.io/en/1.2/demonstrations.html#access-the-full-comsol-api
[`Model.evaluate()`]: https://mph.readthedocs.io/en/1.2/api/mph.Model.html#mph.Model.evaluate

**Comment by bva99** [comment]
There is a way of doing it, indeed.

Assuming that the tag of the global evaluation group is "gev1", then you can do:
```python
import mph
import numpy as np
client = mph.start()
model = client.load("your_file.mph")

eval_results = np.array(model.java.result().numerical("gev1").getReal())
```

The function `model.java.result().numerical("gev1").getReal()` evaluates the global evaluation if it hasn't done so already, then gives a reference to a Java array with the (real part) of the results. 

There are three ways to learn how to directly work with the Java API:
- Looking at the Programming Reference Manual that john-hen showed in his post
- Debugging in Python and looking at `._dir_()`
- Programming in COMSOL itself and learning there how to use the Java API (with, for example, autocompletion and Recording changes done on the GUI and seeing the code it generates).
